WebUI/Pages/Friends.razor

```diff
@@ -1,7 +1,121 @@
+{
+    <h3>Loading...</h3>
+}
+else
+{
+    <div>
+        <button class="tab alert @(friendsTab ? "alert-primary" : "")" @onclick="() => SwitchToFriendsTab(true)">Your Friends</button>
+        <button class="tab alert @(friendsTab ? "" : "alert-primary")" @onclick="() => SwitchToFriendsTab(false)">Add a Friend</button>
+    </div>
+
+    @if (friendsTab)
+    {
+        if (friends.Count == 0)
+        {
+            <h4>You have no friends yet. Go to the "Add a Friend" tab to find new friends!</h4>
+        }
+        else
+        {
+            <div class="d-inline-flex flex-wrap">
+                @foreach (var friend in friends)
+                {
+                    <UserCard OnClick="@(() => RemoveFriend(friend.Id))"
+                              User="friend"
+                              ButtonText="Remove Friend" />
+                }
+            </div>
+        }
+    }
+    else
+    {
+        <input id="Search" @onchange="@LookForUser" placeholder="Enter the username you want to search for..." />
+
+        if (possibleFriends.Count == 0)
+        {
+            <h4>There are no available friends at the moment :(</h4>
+        }
+        else
+        {
+            <div class="d-inline-flex flex-wrap">
+                @foreach (var possibleFriend in possibleFriends)
+                {
+                    <UserCard OnClick="@(() => AddFriend(possibleFriend.Id))"
+                              User="possibleFriend"
+                              ButtonText="Add friend" />
+                }
+            </div>
+        }
+    }
+}
+
 @code {
+    [CascadingParameter]
+    private Task<AuthenticationState> AuthenticationState { get; set; }
+    private AuthenticationState authenticationState;
+    private ApplicationUser currentUser;
+    private List<ApplicationUser> friends;
+    private List<ApplicationUser> possibleFriends = new List<ApplicationUser> { };
+    private bool friendsTab = true;
+
+    protected override async Task OnInitializedAsync()
+    {
+        authenticationState = await AuthenticationState;
+        currentUser = await userService.GetUser(authenticationState.User.Identity.Name);
+        friends = await userService.GetFriends(currentUser);
+        possibleFriends = await GetPossibleFriends();
+    }
+
+    private async Task<List<ApplicationUser>> GetPossibleFriends()
+    {
+        var users = await userService.GetUsers();
+        var friendsIds = friends.Select(u => u.Id).ToList();
+
+        return users.Where(u => u.Id != currentUser.Id && !friendsIds.Contains(u.Id)).ToList();
+    }
+
+    private void SwitchToFriendsTab(bool isFriendsTab)
+    {
+        friendsTab = isFriendsTab;
+    }
+
+    private async void AddFriend(string userId)
+    {
+        await userService.CreateFriendship(currentUser.Id, userId);
+        friends = await userService.GetFriends(currentUser);
+        possibleFriends = await GetPossibleFriends();
+        StateHasChanged();
```

Comment:
Extracting to a method actually worked for me (at least I haven't noticed any issues). I extracted these 3 lines into `RefreshFriends()`

Reply:
Hmmmm. I don't know then. Maybe you can just do the change and commit to this branch?